<!DOCTYPE html>
<html lang="en">
<head>

        <title>Tech Notes</title>
        <meta charset="utf-8" />
        <link href="http://xiaojianfeng.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Tech Notes Full Atom Feed" />
        <link href="http://xiaojianfeng.github.io/feeds/bioinf.atom.xml" type="application/atom+xml" rel="alternate" title="Tech Notes Categories Atom Feed" />


        <!-- Mobile viewport optimized: j.mp/bplateviewport -->
        <meta name="viewport" content="width=device-width,initial-scale=1, maximum-scale=1">

        <link rel="stylesheet" type="text/css" href="http://xiaojianfeng.github.io/theme/gumby.css" />
        <link rel="stylesheet" type="text/css" href="http://xiaojianfeng.github.io/theme/style.css" />
        <link rel="stylesheet" type="text/css" href="http://xiaojianfeng.github.io/theme/pygment.css" />

        <script src="http://xiaojianfeng.github.io/theme/js/libs/modernizr-2.6.2.min.js"></script>




</head>

<body id="index" class="home">


    <div class="container">

        <div class="row">

          <header id="banner" class="body">
                  <h1><a href="http://xiaojianfeng.github.io/">Tech Notes <strong></strong></a></h1>
          </header><!-- /#banner -->

            <div id="navigation" class="navbar row">
              <a href="#" gumby-trigger="#navigation &gt; ul" class="toggle"><i class="icon-menu"></i></a>
             
              <ul class="columns">
                <li><a href="http://xiaojianfeng.github.io/">Home</a></li>


              </ul>
            </div>

<section id="content">
<h2>Articles in the bioinf category</h2>


   <div class="row">


        <div class="eleven columns">
                 <ol id="post-list">
                        <li><article class="hentry">
                                <header> <h2 class="entry-title"><a href="http://xiaojianfeng.github.io/bioinf/common-mistakes-in-bioinformatics.html" rel="bookmark" title="Permalink to common mistakes in bioinformatics">common mistakes in bioinformatics</a></h2> </header>
                                <footer class="post-info">
                                    <abbr class="published" title="2015-03-14T16:03:00+08:00"> Sat 14 March 2015 </abbr>
                                    <address class="vcard author">By 
                                    <a class="url fn" href="http://xiaojianfeng.github.io/author/xiao-jianfeng.html"> Xiao Jianfeng</a>
                                    </address>
                                </footer><!-- /.post-info -->
                                <div class="entry-content"> <p>This nice thread(https://www.biostars.org/p/7126/) summarized a lot of common mistakes people (including me, of course) might make in daily work.</p>
<div class="highlight"><pre><span class="o">*</span> <span class="nt">forget</span> <span class="nt">to</span> <span class="nt">check</span> <span class="nt">both</span> <span class="nt">strands</span>
<span class="o">*</span> <span class="nt">generate</span> <span class="nt">random</span> <span class="nt">genomic</span> <span class="nt">sites</span> <span class="nt">without</span> <span class="nt">avoiding</span> <span class="nt">masked</span> <span class="o">(</span><span class="nt">NNN</span><span class="o">)</span> <span class="nt">gaps</span>
<span class="o">*</span> <span class="nt">confuse</span> <span class="nt">genome</span> <span class="nt">freezes</span> <span class="nt">and</span> <span class="nt">even</span> <span class="nt">species</span>
<span class="o">*</span> <span class="nt">a</span> <span class="nt">lot</span> <span class="nt">of</span> <span class="nt">off-by-one</span> <span class="nt">errors</span> <span class="nt">due</span> <span class="nt">to</span> <span class="nt">switching</span> <span class="nt">between</span> <span class="nt">formats</span><span class="o">:</span> <span class="nt">Bed</span> <span class="nt">is</span> <span class="nt">0</span> <span class="nt">based</span><span class="o">;</span>  <span class="nt">GFF</span><span class="o">/</span><span class="nt">GTF</span> <span class="nt">are</span> <span class="nt">1-based</span>  
<span class="o">*</span> <span class="nt">Trusting</span> <span class="nt">that</span> <span class="nt">a</span> <span class="nt">downloaded</span> <span class="nt">file</span> <span class="nt">is</span> <span class="nt">actually</span> <span class="nt">fully</span> <span class="nt">downloaded</span>
<span class="o">*</span> <span class="nt">Trusting</span> <span class="nt">that</span> <span class="nt">an</span> <span class="nt">aligner</span> <span class="nt">will</span> <span class="nt">accept</span> <span class="nt">a</span> <span class="nt">list</span> <span class="nt">of</span> <span class="nt">query</span> <span class="nt">files</span> <span class="nt">instead</span> <span class="nt">of</span> <span class="nt">just</span> <span class="nt">taking</span> <span class="nt">the</span> <span class="nt">first</span> <span class="nt">and</span> <span class="nt">ignoring</span> <span class="nt">the</span> <span class="nt">rest</span> <span class="o">(</span><span class="nt">quiz</span><span class="o">:</span> <span class="nt">which</span> <span class="nt">ones</span> <span class="nt">am</span> <span class="nt">i</span> <span class="nt">talking</span> <span class="nt">about</span><span class="o">?)</span>
<span class="o">*</span> <span class="nt">Assuming</span> <span class="nt">that</span> <span class="nt">the</span> <span class="nt">quality</span> <span class="nt">scores</span> <span class="nt">in</span> <span class="nt">a</span> <span class="nt">FASTQ</span> <span class="nt">file</span> <span class="nt">are</span> <span class="nt">from</span> <span class="nt">a</span> <span class="nt">great</span> <span class="nt">Sanger-encoded</span> <span class="nt">run</span> <span class="nt">instead</span> <span class="nt">of</span> <span class="nt">a</span> <span class="nt">very</span> <span class="nt">poor</span> <span class="nt">Illumina-1</span><span class="nc">.3</span> <span class="nt">run</span>
<span class="o">*</span> <span class="nt">Assuming</span> <span class="nt">chr1</span> <span class="nt">is</span> <span class="nt">followed</span> <span class="nt">by</span> <span class="nt">chr2</span> <span class="nt">not</span> <span class="nt">chr10</span>
<span class="o">*</span> <span class="nt">off-by-one</span> <span class="nt">errors</span>
<span class="o">*</span> <span class="nt">regex</span> <span class="nt">errors</span>
<span class="o">*</span> <span class="nt">parsing</span> <span class="nt">a</span> <span class="nt">complex</span> <span class="nt">alignment</span><span class="o">/</span><span class="nt">file</span> <span class="nt">format</span> <span class="nt">incorrectly</span> <span class="o">(</span><span class="nt">e</span><span class="nc">.g</span><span class="o">.</span> <span class="nt">BLAST</span> <span class="nt">or</span> <span class="nt">GenBank</span><span class="o">,</span> <span class="nt">probably</span> <span class="nt">the</span> <span class="nt">original</span> <span class="nt">rationale</span> <span class="nt">for</span> <span class="nt">developing</span> <span class="nt">BioPerl</span><span class="o">)</span>
<span class="o">*</span> <span class="nt">failing</span> <span class="nt">to</span> <span class="nt">account</span> <span class="nt">for</span> <span class="nt">strand</span>
<span class="o">*</span> <span class="nt">failing</span> <span class="nt">to</span> <span class="nt">revcomp</span> <span class="nt">sequences</span>
<span class="o">*</span> <span class="nt">failing</span> <span class="nt">to</span> <span class="nt">account</span> <span class="nt">for</span> <span class="nt">the</span> <span class="nt">last</span> <span class="nt">element</span> <span class="nt">in</span> <span class="nt">a</span> <span class="nt">file</span> <span class="o">(</span><span class="nt">because</span> <span class="nt">of</span> <span class="nt">a</span> <span class="nt">improper</span> <span class="nt">loop</span> <span class="nt">condition</span> <span class="nt">or</span> <span class="nt">no</span> <span class="nt">EOL</span> <span class="nt">character</span> <span class="nt">on</span> <span class="nt">last</span> <span class="nt">line</span><span class="o">)</span>
<span class="o">*</span> <span class="nt">failing</span> <span class="nt">to</span> <span class="nt">account</span> <span class="nt">for</span> <span class="nt">OS</span> <span class="nt">dependent</span> <span class="nt">line</span> <span class="nt">breaks</span>
<span class="o">*</span> <span class="nt">using</span> <span class="nt">the</span> <span class="nt">wrong</span> <span class="nt">assembly</span><span class="o">/</span><span class="nt">annotation</span><span class="o">/</span><span class="nt">release</span>
<span class="o">*</span> <span class="nt">using</span> <span class="nt">the</span> <span class="nt">wrong</span> <span class="nt">genome</span> <span class="nt">coordinate</span> <span class="nt">system</span>
<span class="o">*</span> <span class="nt">using</span> <span class="nt">the</span> <span class="nt">wrong</span> <span class="nt">file</span> <span class="o">(</span><span class="nt">multiple</span> <span class="nt">versions</span><span class="o">,</span> <span class="nt">version</span> <span class="nt">skew</span><span class="o">)</span>
<span class="o">*</span> <span class="nt">failing</span> <span class="nt">to</span> <span class="nt">account</span> <span class="nt">for</span> <span class="nt">nested</span><span class="o">/</span><span class="nt">intercalated</span> <span class="nt">annotation</span> <span class="nt">features</span> <span class="o">(</span><span class="nt">e</span><span class="nc">.g</span><span class="o">.</span> <span class="nt">genes</span><span class="o">)</span>
<span class="o">*</span> <span class="nt">assuming</span> <span class="nt">all</span> <span class="nt">jobs</span> <span class="nt">have</span> <span class="nt">completed</span> <span class="nt">on</span> <span class="nt">a</span> <span class="nt">cluster</span>
<span class="o">*</span> <span class="nt">deleting</span> <span class="nt">files</span>
<span class="o">*</span> <span class="nt">not</span> <span class="nt">randomizing</span> <span class="nt">your</span> <span class="nt">data</span> <span class="nt">properly</span>
<span class="o">*</span> <span class="nt">improper</span> <span class="nt">use</span> <span class="nt">of</span> <span class="nt">statistical</span> <span class="nt">tests</span>
<span class="o">*</span> <span class="nt">not</span> <span class="nt">documenting</span> <span class="nt">methods</span> <span class="nt">fully</span> <span class="o">(</span><span class="nt">to</span> <span class="nt">check</span> <span class="nt">and</span> <span class="nt">correct</span> <span class="nt">all</span> <span class="nt">of</span> <span class="nt">the</span> <span class="nt">above</span><span class="o">)</span>
<span class="o">*</span> <span class="nt">Run</span> <span class="nt">a</span> <span class="nt">batch</span> <span class="nt">BLAST</span> <span class="nt">job</span> <span class="nt">and</span> <span class="nt">forgetting</span> <span class="nt">to</span> <span class="nt">put</span> <span class="nt">the</span> <span class="s2">&quot;-o something.out&quot;</span> <span class="nt">option</span><span class="o">.</span> <span class="nt">Then</span> <span class="nt">switching</span> <span class="nt">off</span> <span class="nt">the</span> <span class="nt">monitor</span> <span class="nt">and</span> <span class="nt">coming</span> <span class="nt">the</span> <span class="nt">next</span> <span class="nt">day</span> <span class="nt">to</span> <span class="nt">see</span> <span class="nt">a</span> <span class="nt">bunch</span> <span class="nt">of</span> <span class="nt">characters</span> <span class="nt">in</span> <span class="nt">my</span> <span class="nt">terminal</span>
<span class="o">*</span> <span class="s2">&quot;tar -zxvf&quot;</span> <span class="nt">without</span> <span class="nt">checking</span> <span class="nt">the</span> <span class="nt">tar</span> <span class="nt">file</span> <span class="nt">before</span><span class="o">,</span> <span class="nt">I</span> <span class="nt">have</span> <span class="nt">decompressed</span> <span class="nt">thousands</span> <span class="nt">of</span> <span class="nt">files</span> <span class="nt">in</span> <span class="nt">my</span> <span class="nt">current</span> <span class="nt">directory</span> <span class="nt">assuming</span> <span class="nt">they</span> <span class="nt">came</span> <span class="nt">in</span> <span class="nt">their</span> <span class="nt">own</span> <span class="nt">folder</span><span class="o">.</span>
</pre></div>


<p>Another very important one is: gene name might be hacked by MS Excel. </p>
<p>There are 11 human gene names in March: MARCH1 to MARCH11, and one in December: DEC1. And excel will treat them as date!</p>
<p>So make sure to ask excel to open data as text when working with gene names.</p>
<p>Please refer to this paper for more info.</p>
<p>1.Zeeberg, B. R. et al. Mistaken Identifiers: Gene name errors can be introduced inadvertently when using Excel in bioinformatics. BMC Bioinformatics 5, 80 (2004).  http://www.biomedcentral.com/1471-2105/5/80</p> </div><!-- /.entry-content -->

                                <div class="medium primary btn"><a href="http://xiaojianfeng.github.io/bioinf/common-mistakes-in-bioinformatics.html" rel="bookmark" title="Permalink to common mistakes in bioinformatics">Read more <i class="icon-arrow-right"></i></a></div>



                                <div class="row tag-row">
                                </div>



                        </article></li>
                        <li><article class="hentry">
                                <header> <h2 class="entry-title"><a href="http://xiaojianfeng.github.io/bioinf/convert-bam-file-to-fastq.html" rel="bookmark" title="Permalink to convert bam file to fastq">convert bam file to fastq</a></h2> </header>
                                <footer class="post-info">
                                    <abbr class="published" title="2014-08-06T20:16:55+08:00"> Wed 06 August 2014 </abbr>
                                    <address class="vcard author">By 
                                    <a class="url fn" href="http://xiaojianfeng.github.io/author/xiao-jianfeng.html"> Xiao Jianfeng</a>
                                    </address>
                                </footer><!-- /.post-info -->
                                <div class="entry-content"> <p>After having struggled for a while to find a convenient tool to extract .fastq file to .bam file, I decided to write one.</p>
<p>The problems with current tools are:</p>
<ul>
<li>Couldn't deal with reads having multiple match in .bam file.</li>
<li>Some tools require the .bam to be sorted by name first.</li>
</ul>
<p>In my program (https://github.com/XiaoJianfeng/ibl/blob/master/util/bam2fq.sh), Sqlite3 is used to do the heavy job. Below is the details of how to do it:</p>
<ol>
<li>
<p>"samtools view" read .bam file and output to stdout.</p>
</li>
<li>
<p>Sqlite3 create a table fastq with 4 columns (qname, flag, seq, qual), and import data from stdout to the table with sqlite3 command ".import".
CREATE TABLE fastq (qname, flag, seq, qual);</p>
<p>select date()||" "||time()||" import begin";<br />
.separator "\t"<br />
.import temp.$$ fastq 
CREATE INDEX qname_idx on fastq (qname);   </p>
</li>
<li>
<p>Sqlite3 creat two tables, "left" and "right", and get qname of read_1 and read_2 to these two tables separately. whether a read is the first read or second read is identified by checking the flag. 0x40 is set for first read, while 0x80 is set for second read.</p>
<p>CREATE TABLE left as select qname from fastq where flag &amp; 64 == 64;<br />
CREATE TABLE right as select qname from fastq where flag &amp; 128 == 128;  </p>
</li>
<li>
<p>select reads from table "fastq" that are the first in the pair, and has corresponding mate pair in table "right"; and select reads from table "fastq" that are the second in the pair, and has corresponding mate pair in table "left". "\n" is used as separator between columns, and the output is redirected to .fastq file.</p>
<p>.output ${left_file}<br />
.separator "\n"<br />
select distinct "@"||right.qname||"/1", seq, '+', qual from right join fastq on fastq.qname == right.qname where flag&amp;64 == 64 order by right.qname;  </p>
<p>.output ${right_file}<br />
.separator "\n"<br />
select distinct "@"||left.qname||"/2", seq, '+', qual from left join fastq on fastq.qname == left.qname where flag&amp;128 == 128 order by left.qname;  </p>
</li>
<li>
<p>job done.</p>
</li>
</ol>
<p>With my program, the .bam file doesn't need to be sorted beforehand.</p>
<p>It has one limition: the output .fastq files are not compressed. I will remove this limition later when I have time. But please rest assured it works perfectly well at the moment.</p>
<h2>[update 2014-08-11 22:33:09]:</h2>
<h3>fixed a bug regarding identification of first and second read in a pair by flags</h3>
<p>the primary alignment should have flag &amp; 0x900 == 0, first read has flag &amp; 0x40 == 0x40, and second read has flag &amp; 0x80 == 0x80</p>
<p>so primary alignment of first read should be flag &amp; 0x940 (ie. 2368) == 0x40, and for second read flag &amp; 0x980 (ie. 2432) == 0x80</p>
<p>CREATE TABLE left as select qname from fastq where flag &amp; 2368 == 64; <br />
   CREATE TABLE right as select qname from fastq where flag &amp; 2432 == 128;       </p>
<h3>bam2fq.sh is updated to support gziped output.</h3>
<p>details of the implementation:</p>
<ul>
<li>
<p>first creat named pipe</p>
<p>mkfifo tmp_output</p>
</li>
<li>
<p>use gzip to compress the content read from named pipe, and put the process into background</p>
<p>gzip -c &lt; tmp_output &gt; output.fq.gz &amp;</p>
</li>
<li>
<p>in sqlite3, output to to named pipe</p>
<p>.output tmp_output
[SELECT ...]</p>
</li>
<li>
<p>it is possible to set .output to a pipe from within sqlite3,</p>
<p>.output '| gzip -c &gt; output.fq.gz'
[SELECT ...]</p>
</li>
</ul>
<h2>[2014-08-12 21:14:52]</h2>
<h3>the task is more complicated than I have expected. here is another bug</h3>
<p>The sequences in .bam file for two reads of a pair are on the same strand. However, for 
paired-end reads, one read is on forward stand, and the other one is on reverse strand.
So one of them should be mapped to another strand (reverse complement).</p>
<p>In bam2fq.sh, the first reads are extracted from .bam unchanged. The second reads are all converted to reverse complement with command "tr 'ATGC' 'TACG' | rev".</p>
<p><a href="https://github.com/XiaoJianfeng/ibl/blob/master/util/bam2fq.sh">bam2fq.sh</a> is updated.</p> </div><!-- /.entry-content -->

                                <div class="medium primary btn"><a href="http://xiaojianfeng.github.io/bioinf/convert-bam-file-to-fastq.html" rel="bookmark" title="Permalink to convert bam file to fastq">Read more <i class="icon-arrow-right"></i></a></div>



                                <div class="row tag-row">
                                </div>



                        </article></li>
                </ol><!-- /#posts-list -->

        </div><!-- /.eleven.columns -->
        
<div class="three columns">

<h4>Pages</h4>

 <ul>
  </ul>

<h4>Categories</h4>
<ul>
		<li><a href="http://xiaojianfeng.github.io/bioinf/index.html">bioinf</a></li>
		<li><a href="http://xiaojianfeng.github.io/clinical/index.html">clinical</a></li>
		<li><a href="http://xiaojianfeng.github.io/code/index.html">code</a></li>
		<li><a href="http://xiaojianfeng.github.io/general/index.html">general</a></li>
		<li><a href="http://xiaojianfeng.github.io/mac/index.html">mac</a></li>
		<li><a href="http://xiaojianfeng.github.io/utils/index.html">utils</a></li>
</ul>


<h4>Tags</h4>
	<ul>
	    <li class="tag-4"><a href="http://xiaojianfeng.github.io/tag/shell.html">shell</a></li>
	    <li class="tag-4"><a href="http://xiaojianfeng.github.io/tag/mutation.html">mutation</a></li>
	    <li class="tag-4"><a href="http://xiaojianfeng.github.io/tag/mac.html">mac</a></li>
</ul>


<nav class="widget">
  <h4>Social</h4>
  <ul>
    <li><a href="https://github.com/xiaojianfeng">github</a></li>
    <li><a href="#">Another social link</a></li>
  </ul>
</nav>

</div> </div><!-- /.row -->

<p class="paginator">
    Page 1 / 1
</p>
</section><!-- /#content -->

       </div><!-- /.row -->
    </div><!-- /.container -->


       <div class="container.nopad bg">

    
        <footer id="credits" class="row">
          <div class="seven columns left-center">

                   <address id="about" class="vcard body">
                    Proudly powered by <a href="http://getpelican.com/">Pelican</a>,
                    which takes great advantage of <a href="http://python.org">Python</a>.
                    <br />
                    Based on the <a target="_blank" href="http://gumbyframework.com">Gumby Framework</a>
                    </address>
          </div>


          <div class="seven columns">
            <div class="row">
              <ul class="socbtns">





              </ul>
            </div>
          </div>
        </footer>

    </div>


  <script src="http://xiaojianfeng.github.io/theme/js/libs/jquery-1.9.1.min.js"></script>
  <script src="http://xiaojianfeng.github.io/theme/js/libs/gumby.min.js"></script>
  <script src="http://xiaojianfeng.github.io/theme/js/plugins.js"></script>
</body>
</html>